File: App/src/Adaptive.ReactiveTrader.Client.Domain/Transport/Connection.cs
using System;
using System.Reactive;
using System.Reactive.Disposables;
using System.Reactive.Linq;
using System.Reactive.Subjects;
using Adaptive.ReactiveTrader.Shared;
using Adaptive.ReactiveTrader.Shared.Extensions;
using log4net;
using Microsoft.AspNet.SignalR.Client;

namespace Adaptive.ReactiveTrader.Client.Domain.Transport
{
    /// <summary>
    /// This represents a single SignalR connection.
    /// The <see cref="ConnectionProvider"/> creates connections and is responsible for creating new one when a connection is closed.
    /// </summary>
    internal class Connection : IConnection
    {
        private static readonly ILog Log = LogManager.GetLogger(typeof(Connection));

        private readonly ISubject<ConnectionInfo> _statusStream;
        private readonly HubConnection _hubConnection;

        private bool _initialized;

        public Connection(string address, string username)
        {
            _statusStream = new BehaviorSubject<ConnectionInfo>(new ConnectionInfo(ConnectionStatus.Uninitialized, address));
            Address = address;
            _hubConnection = new HubConnection(address);
            _hubConnection.Headers.Add(ServiceConstants.Server.UsernameHeader, username);
            CreateStatus().Subscribe(
                s => _statusStream.OnNext(new ConnectionInfo(s, address)),
                _statusStream.OnError,
                _statusStream.OnCompleted);
            _hubConnection.Error += exception => Log.Error("There was a connection error with " + address, exception);

            BlotterHubProxy = _hubConnection.CreateHubProxy(ServiceConstants.Server.BlotterHub);
            ExecutionHubProxy = _hubConnection.CreateHubProxy(ServiceConstants.Server.ExecutionHub);
            PricingHubProxy = _hubConnection.CreateHubProxy(ServiceConstants.Server.PricingHub);
            ReferenceDataHubProxy = _hubConnection.CreateHubProxy(ServiceConstants.Server.ReferenceDataHub);
        }

        public IObservable<Unit> Initialize()
        {
            if (_initialized)
            {
                throw new InvalidOperationException("Connection has already been initialized");
            }
            _initialized = true;

            return Observable.Create<Unit>(async observer =>
            {
                _statusStream.OnNext(new ConnectionInfo(ConnectionStatus.Connecting, Address)); 

                try
                {
                    Log.InfoFormat("Connecting to {0}", Address);
                    await _hubConnection.Start();
                    _statusStream.OnNext(new ConnectionInfo(ConnectionStatus.Connected, Address));
                    observer.OnNext(Unit.Default);
                }
                catch (Exception e)
                {
                    Log.Error("An error occured when starting SignalR connection", e);
                    observer.OnError(e);
                }

                return Disposable.Create(() =>
                {
                    try
                    {
                        Log.Info("Stoping connection...");
                        _hubConnection.Stop();
                        Log.Info("Connection stopped");
                    }
                    catch (Exception e)
                    {
                        // we must never throw in a disposable
                        Log.Error("An error occured while stoping connection", e);
                    }
                });
            })
            .Publish()
            .RefCount();
        } 

        private IObservable<ConnectionStatus> CreateStatus()
        {
            var closed = Observable.FromEvent(h => _hubConnection.Closed += h, h => _hubConnection.Closed -= h).Select(_ => ConnectionStatus.Closed);
            var connectionSlow = Observable.FromEvent(h => _hubConnection.ConnectionSlow += h, h => _hubConnection.ConnectionSlow -= h).Select(_ => ConnectionStatus.ConnectionSlow);
            var reconnected = Observable.FromEvent(h => _hubConnection.Reconnected += h, h => _hubConnection.Reconnected -= h).Select(_ => ConnectionStatus.Reconnected);
            var reconnecting = Observable.FromEvent(h => _hubConnection.Reconnecting += h, h => _hubConnection.Reconnecting -= h).Select(_ => ConnectionStatus.Reconnecting);
            return Observable.Merge(closed, connectionSlow, reconnected, reconnecting)
                .TakeUntilInclusive(status => status == ConnectionStatus.Closed); // complete when the connection is closed (it's terminal, SignalR will not attempt to reconnect anymore)
        }

        public IObservable<ConnectionInfo> StatusStream
        {
            get { return _statusStream; }
        }

        public string Address { get; private set; }

        public IHubProxy ReferenceDataHubProxy { get; private set; }
        public IHubProxy PricingHubProxy { get; private set; }
        public IHubProxy ExecutionHubProxy { get; private set; }
        public IHubProxy BlotterHubProxy { get; private set; }

        public override string ToString()
        {
            return string.Format("Address: {0}", Address);
        }
    }
}

File: App/src/Adaptive.ReactiveTrader.Client.JS/Transport/Connection.ts
/// <reference path="../typings/signalr/signalr.d.ts"/> 
/// <reference path="../typings/rx.js/rx.d.ts"/>

class Connection implements IConnection {
    private _status: Rx.Subject<ConnectionInfo>;
    private _hubConnection: HubConnection;
    private _initialized: boolean;
    private _address: string;
    private _referenceDataHubProxy: HubProxy;
    private _pricingHubProxy: HubProxy;
    private _executionHubProxy: HubProxy;
    private _blotterHubProxy: HubProxy;
    private _allPrices: Rx.Subject<PriceDto>;
    private _currencyPairUpdates: Rx.Subject<CurrencyPairUpdateDto>;

    constructor(address: string, username: string) {
        //var initialConnectionInfo = new ConnectionInfo(ConnectionStatus.Uninitialized, address);
        this._status = new Rx.Subject<ConnectionInfo>();
        this._address = address;
        this._hubConnection = $.hubConnection("http://localhost:8080/signalr");

        // TODO set username header 
        this._hubConnection
            .disconnected(() => this.changeStatus(ConnectionStatus.Closed))
            .connectionSlow(() => this.changeStatus(ConnectionStatus.ConnectionSlow))
            .reconnected(() => this.changeStatus(ConnectionStatus.Reconnected))
            .reconnecting(() => this.changeStatus(ConnectionStatus.Reconnecting))
            .error(error => console.log(error));

        this._referenceDataHubProxy = this._hubConnection.createHubProxy("ReferenceDataHub");
        this._blotterHubProxy = this._hubConnection.createHubProxy("BlotterHub");
        this._executionHubProxy = this._hubConnection.createHubProxy("ExecutionHub");
        this._pricingHubProxy = this._hubConnection.createHubProxy("PricingHub");

        this.installListeners();
    }

    public initialize(): Rx.Observable<{}> {

        return Rx.Observable.create<{}>(observer => {
            this.changeStatus(ConnectionStatus.Connecting);

            console.log("Connecting to " + this._address + "...");
            this._hubConnection.start()
                .done(()=> {
                    this.changeStatus(ConnectionStatus.Connected);
                    observer.onNext(true);
                    console.log("Connected to " + this._address + ".");
                })
                .fail(()=> {
                    this.changeStatus(ConnectionStatus.Closed);
                    var error = "An error occured when starting SignalR connection.";
                    console.log(error);
                    observer.onError(error);
                });
            
            return Rx.Disposable.create(() => {
                console.log("Stoping connection...");
                this._hubConnection.stop();
                console.log("Connection stopped.");
            });
        });
        //.publish()
        //.refCount();
    }

    public changeStatus(newStatus: ConnectionStatus): void {
        this._status.onNext(new ConnectionInfo(newStatus, this.address));
    }

    public get status(): Rx.Observable<ConnectionInfo> {
        return this._status;
    }

    public get address(): string {
        return this._address;
    }

    public get referenceDataHubProxy(): HubProxy {
        return this._referenceDataHubProxy;
    }

    public get pricingHubProxy(): HubProxy {
        return this._pricingHubProxy;
    }

    public get executionHubProxy(): HubProxy {
        return this._executionHubProxy;
    }

    public get blotterHubProxy(): HubProxy {
        return this._blotterHubProxy;
    }

    public get allPrices(): Rx.Observable<PriceDto> {
        return this._allPrices;
    }

    public get currencyPairUpdates(): Rx.Observable<CurrencyPairUpdateDto> {
        return this._currencyPairUpdates;
    }

    private installListeners() {
        this._allPrices = new Rx.Subject<PriceDto>();
        this._currencyPairUpdates = new Rx.Subject<CurrencyPairUpdateDto>();

        this._pricingHubProxy.on("OnNewPrice", price=> this._allPrices.onNext(price));
        this._referenceDataHubProxy.on("OnCurrencyPairUpdate", currencyPairs=> this._currencyPairUpdates.onNext(currencyPairs));
    }
}
